client/app/src/components/addressbook/addressbook.controller.js

```diff
@@ -63,6 +63,55 @@
       }
     }
 
+    self.addressBookAddContactRecord = function(name, address, callback, isRename, suppressNotice) {
+      self.getContacts();
+      if (self.trim(name) == "") {
+        self.showToast('This Contact Name is not valid', name, true);
+        return;
+      }
+      if (self.trim(address) == "" || ! self.isAddress(address)) {
+        self.showToast('This Contact Address is not valid', address, true);
+        return;
+      }
+
+      var newContact = { name: name, address: address };
+      if (!isRename && self.addressExists(address)) {
+        self.showToast('A Contact with this Address already exists', address, true);
+        return;
+      }
+      if (self.contactExists(name)) {
+        self.showToast('A Contact with this Name already exists', name, true);
+        return;
+      }
+
+      var knownAccounts = accountService.loadAllAccounts().reduce( (all, account) => {
+        if (account.virtual) {
+          all.push(account);
+        }
+        return all;
+      }, []);
+
+      if (existsIn(knownAccounts.map( a => a.address ), address)) {
+        self.showToast('This Address is already taken, please add a different one', address, true);
+        return;
+      }
+      if (existsIn(knownAccounts.map( a => a.username ), name)) {
+        self.showToast('This Name is already taken, please use a different one', name, true);
+        return;
+      }
+
+      self.contacts.push(newContact);
+      self.save();
+      if (!suppressNotice) {
+        self.showToast('Contact successfully added', name, false);
+      }
+      if (typeof callback === 'function') {
+        callback();
+      }
+
+      return true;
+    }
```

Comment:
Can this method be broken into smaller unit-testable chunks? Do we need to return false when an error comes up above?